### PR TITLE
New font icons, Modernizr, and A Font Garde

### PR DESCRIPTION
**Pull request by clrux**
## Accessible icon font and fallbacks

Our pattern library previously used SVG paths for icons. While this method was accessible, it required a bunch of markup and the SVG paths were difficult to scale alongside text using font size. Going back even further, we used CSS-based icon fonts that relied on the `:before` pseudo attribute, and while incredibly easy to use, broke if `@font-face` wasn't available or if the font icons didn't load. Users would end up with unhelpful glyph boxes. This was especially painful in instances where the icon was the only indication of the action.

### A Font Garde + Modernizr

The approach we're using here comes from The Filament Group's blog post ["Bulletproof Accessible Icon Fonts"](https://www.filamentgroup.com/lab/bulletproof_icon_fonts.html) and uses feature detection plus a nifty piece of JavaScript to check if the font icons load. If they do, a class is appended to the `html` of the document and the icon fonts render using that class. If the test fails, fallbacks are displayed instead.

#### Simplified markup

Rather than referencing an SVG and its definitions file, we now use one of the following fallback options:

##### Critical icons that fallback to text
```
<span class="icon-fallback-text">
    <span class="icon icon-video-camera" aria-hidden="true"></span>
    <span class="text">Video</span>
</span>
```

##### Critical icons that fallback to an image
```
<span class="icon-fallback-img">
    <span class="icon icon-video-camera" aria-hidden="true"></span>
    <span class="text">Video</span>
</span>
```

#### When font icons are available

When the tests pass, the name of the icon font (`.edx-icons` in our case) is appended to the `html` of the document, alongside a couple of Modernizr check classnames that are used by the `afontgarde.js` file indirectly.

We then style our icon fonts to rely on this class if it's present:

```
.edx-icons .icon {
    font-family: $icon-font;
    ...
}

.edx-icons .icon-linkedin-square:before {
    content: "\e600";
}
...
```

#### When font icons are _not_ available

If the font icon doesn't load, if `@font-face` isn't available in the browser, or if the user chooses not to let the site use its available stylesheets, we don't want random, unhelpful glyph boxes. Since our icon fonts look for the appended `html` class they won't load, giving us a chance to create fallbacks.

##### Critical icons that fallback to text

Just the text in `text` is displayed. The icon itself is hidden.

##### Critical icons that fallback to an image

We create separate rules to use CSS background images instead, like the following:

```
.icon-linkedin-square {
    background: url('path-to-image') no-repeat;
}
```

## Notable changes

There were a number of changes in this update. A summary of them is as follows:

* removed the SVG icons and their definitions file
* added a new icon font set (generated with the same settings file at Icomoon)
* added a new `afontgarde` package to Bower
* added a new `modernizr` package to Bower
* added both of the above to the compile and load
* updated all of the example icons
* updated the sizing script to update font size

### Post-Icomoon required changes

While we are able to utilize _most_ of the stock icons from Icomoon, we do need to make a few easy changes in order for this process to work.

* Icomoon spits out a `.css` file which I copied and placed into the `_icons.scss` partial.
* `.edx-icons` will need to be added in front of every `.icon` in that same partial
* place any fallbacks at the bottom

## Browser testing

This work was checked in the following browsers:

| Platform | Browser | Version | Passed |
|----------|---------|---------|--------|
| Mac      | Chrome  | 45      | Yes    |
|          | Safari  | 8       | Yes    |
|          | Firefox | 40      | Yes *  |
| Windows  | Chrome  | 45      | Yes    |
|          | Firefox | 40      | Yes *    |
|          | IE      | 10      | - *    |
|          | IE      | 11      | - *    |

_* A note about Firefox: Firefox has a known bug where it doesn't load or download local fonts. So, previewing the fonts with `@font-face` disabled will result in the fallbacks being triggered. You'll need to view this from a server to see it in action.

## Reviewers

- [ ] @talbs (Sass and overall overallness)
- [ ] @frrrances  (second comprehender)

### CODE REVIEW

File: gulp/config.js [hunk 1]
```diff
@@ -35,7 +35,7 @@ module.exports = {
 
         // documentation site
         pldoc_src:              pldoc_src + '/static/sass',
-        pldoc_src_files:        pldoc_src + '/static/sass/**/*.scss'
+        pldoc_src_files:        pldoc_src + '/static/sass/**/*'
```
Comment (talbs) [inline gulp/config.js]: Why was this changed as part of this work?

File: gulp/config.js [hunk 2]
```diff
@@ -59,27 +59,16 @@ module.exports = {
...
-
-        // documentation site
-        pldoc_src:              pldoc_src + '/static/icons',
-        pldoc_src_files:        pldoc_src + '/static/icons/**/*'
-    },
     scripts:                    {
         // pattern library
         src:                    src + '/js',
-        src_files:              src + '/js/*.js',
+        src_files:              src + '/js/**/*',
```
Comment (talbs) [inline gulp/config.js]: Why was this changed as part of this work?

File: gulp/config.js [hunk 3]
```diff
@@ -59,27 +59,16 @@ module.exports = {
...
         dest:                   dest + '/js',
         local:                  local + '/public/js',
 
         // documentation site
         pldoc_src:              pldoc_src + '/static/js',
-        pldoc_src_files:        pldoc_src + '/static/js/**/*.js'
+        pldoc_src_files:        pldoc_src + '/static/js/**/*'
```
Comment (talbs) [inline gulp/config.js]: Why was this changed as part of this work?

File: gulp/tasks/styles.js [hunk 4]
```diff
@@ -2,22 +2,25 @@ var gulp            = require('gulp'),
     autoprefixer    = require('gulp-autoprefixer'),
     browserSync     = require('browser-sync'),
     changed         = require('gulp-changed'),
-    config          = require('../config').styles,
+    config          = require('../config'),
```
Comment (talbs) [inline gulp/tasks/styles.js]: Why was this changed as part of this work?!

File: gulp/tasks/styles.js [hunk 5]
```diff
@@ -2,22 +2,25 @@ var gulp            = require('gulp'),
...
 gulp.task('styles', function () {
-    return gulp.src(config.pldoc_src_files)
+    return gulp.src([
+            config.lib.src + '/afontgarde/afontgarde.css',
+            config.styles.pldoc_src_files
+        ])
         .pipe(sourcemaps.init())
-        .pipe(changed(config.dest)) // ignore unchanged files
-        .pipe(sass(config.settings_develepment))
+        .pipe(changed(config.styles.dest)) // ignore unchanged files
```
Comment (talbs) [inline gulp/tasks/styles.js]: Also, added the config specific object over and over here is redundant and less elegant than the way it was.

File: index.html [hunk 6]
```diff
@@ -62,7 +62,7 @@ <h2 class="pldoc-heading pldoc-tab-heading pldoc-element-preview-title">{{post.t
                             <div class="pldoc-settings">
                                 <div class="pldoc-setting pldoc-setting-sizer">
                                     <label for="iconFontSlider" class="sr-only">Drag to resize icons</label>
-                                    <input id="iconFontSlider" type="range" min="12" max="144" step="12" value="24" role="slider" aria-valuemin="12" aria-valuemax="144" aria-valuenow="24">
+                                    <input id="iconFontSlider" type="range" min="12" max="144" step="12" value="48" role="slider" aria-valuemin="12" aria-valuemax="144" aria-valuenow="48">
```
Comment (talbs) [inline index.html]: Why the value change?

File: pattern-library/fonts/edx-icons/edx-icons.css [hunk 7]
```diff
@@ -0,0 +1,355 @@
+@font-face {
```
Comment (talbs) [inline pattern-library/fonts/edx-icons/edx-icons.css]: Is this file generated from something? Icomoon or A Font Garde?

File: pattern-library/fonts/edx-icons/edx-icons.scss [hunk 8]
```diff
@@ -0,0 +1,110 @@
+$icon-linkedin-square: "\e600";
```
Comment (talbs) [inline pattern-library/fonts/edx-icons/edx-icons.scss]: Is this file generated from something? Icomoon or A Font Garde?

File: pattern-library/fonts/edx-icons/edx-icons.svg [hunk 9]
```diff
@@ -0,0 +1,106 @@
+<?xml version="1.0" standalone="no"?>
```
Comment (talbs) [inline pattern-library/fonts/edx-icons/edx-icons.svg]: Looks like this is generated by IcoMoon, right? Any manual steps/changes you made from the produced file?

File: pattern-library/js/edx-icons.js [hunk 10]
```diff
@@ -0,0 +1,15 @@
+define([], function() {
+
+    AFontGarde('edx-icons', {
+        // Add all of the individual icons you want to check for.
+        // I don't think we need to list them all.
+        glyphs: '\uE621\uE622\uE623',
```
Comment (talbs) [inline pattern-library/js/edx-icons.js]: Does this need to be here or is it demo code?

File: pattern-library/js/edx-icons.js [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+        success: function() {
```
Comment (talbs) [inline pattern-library/js/edx-icons.js]: Your ``console.log`` is showing. Please remove for prod.

File: pattern-library/js/edx-icons.js [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+            console.log('Successfully loaded icon fonts!');
+        },
+        error: function() {
```
Comment (talbs) [inline pattern-library/js/edx-icons.js]: Your ``console.log`` is showing. Please remove for prod.

File: pattern-library/sass/components/_icons.scss [hunk 13]
```diff
@@ -10,5 +10,362 @@
 .icon {
     display: inline-block;
     font-size: 1em;
-    fill: currentColor;
+    color: currentColor;
```
Comment (talbs) [inline pattern-library/sass/components/_icons.scss]: Will this still work? I thought this was SVG element dependent.

File: pattern-library/sass/components/_icons.scss [hunk 14]
```diff
@@ -10,5 +10,362 @@
...
 }
+
+// ----------------------------
+// #ICONS
```
Comment (talbs) [inline pattern-library/sass/components/_icons.scss]: Is this generated by IcoMoon or A Font Garde?

File: pattern-library/sass/utilities/_variables.scss [hunk 15]
```diff
@@ -208,10 +209,18 @@ $open-sans-font-path: '#{$font-path}/OpenSans' !default;
...
+    '#{$edx-icons-font-path}/edx-icons',
+    300,
+    $file-formats: woff woff2 ttf
+);
+
 // typography: config - stacks
 $font-family-sans-serif:    'Open Sans','Helvetica Neue', Helvetica, Arial, sans-serif !default;
 $font-family-serif:         'Georgia', Cambria, 'Times New Roman', Times, serif !default;
 $font-family-monospace:     'Bitstream Vera Sans Mono', Consolas, Courier, monospace !default;
+$icon-font:                 'edx-icons', 'Open Sans', sans-serif;
```
Comment (talbs) [inline pattern-library/sass/utilities/_variables.scss]: Please add a ``!default`` tag in case people want to override this.

**Comment by talbs** [comment]
> While we are able to utilize most of the stock icons from Icomoon, we do need to make a few easy changes in order for this process to work.

> Icomoon spits out a .css file which I copied and placed into the _icons.scss partial.
.edx-icons will need to be added in front of every .icon in that same partial
place any fallbacks at the bottom

If we are not directly including this IcoMoon CSS file, we should not be checking it into the repo (as its being processed/moved to several places).

**Comment by talbs** [comment]
@adding @frrrances to this thread.

**Comment by talbs** [comment]
Before this work is merged, please update the Icon specific guidelines/info wiki (https://github.com/edx/ux-pattern-library/wiki/Styleguide:-Icons) with the new methods of mgmt, inclusion, and use of icons in a UI.

**Comment by talbs** [comment]
UPDATE: I've tried to address some of my own feedback in https://github.com/edx/ux-pattern-library/issues/174, so once that is merged some of the notes here should be taken care of. Note the TODOs listed in that PR.

File: gulp/tasks/styles.js (hunk 4 above)
Comment (clrux) [inline gulp/tasks/styles.js]: When I was pulling in the "bower component" for `afontgarde` I needed to pull the CSS from two places, and so I chose to define them each individually. Now that this is rolled into the main "styles" folder, this is unnecessary.

File: gulp/tasks/styles.js (hunk 5 above)
Comment (clrux) [inline gulp/tasks/styles.js]: It was necessary to get things from two places. But going a different direction.

File: index.html (hunk 6 above)
Comment (clrux) [inline index.html]: I was thinking 24px was too small. What do you think?

File: pattern-library/fonts/edx-icons/edx-icons.scss (hunk 8 above)
Comment (clrux) [inline pattern-library/fonts/edx-icons/edx-icons.scss]: This is spit out from Icomoon, but it's unused at the time. I was thinking it might be nice to have a Sass partial in the chance we wanted to do something... but I'm not sure I can see a purpose. Thoughts?

File: pattern-library/fonts/edx-icons/edx-icons.css (hunk 7 above)
Comment (clrux) [inline pattern-library/fonts/edx-icons/edx-icons.css]: This is the CSS generated from Icomoon. This font family is being loaded in `_variables.scss`, but I decided to leave this here so as to require minimal changes to the vendor file. However... we need to modify this file anyway, so I guess we might as well remove this bit.

File: pattern-library/fonts/edx-icons/edx-icons.svg (hunk 9 above)
Comment (clrux) [inline pattern-library/fonts/edx-icons/edx-icons.svg]: Nope, this one is stock from Icomoon.

File: pattern-library/js/edx-icons.js (hunk 10 above)
Comment (clrux) [inline pattern-library/js/edx-icons.js]: This is needed to actually run A Font Garde. We toss in a few random font icons. It uses these to see if the font is available.

File: pattern-library/js/edx-icons.js (hunk 11 above)
Comment (clrux) [inline pattern-library/js/edx-icons.js]: Psst, your console log is showing.

File: pattern-library/sass/components/_icons.scss (hunk 13 above)
Comment (clrux) [inline pattern-library/sass/components/_icons.scss]: We aren't using SVGs anymore :)

File: pattern-library/sass/components/_icons.scss (hunk 14 above)
Comment (clrux) [inline pattern-library/sass/components/_icons.scss]: I can explain more in detail, but our `_icons.scss` partial was pre-existing. I added icons generated from Icomoon to make it easier to manage. I left the `edx-icons.css` file because I was hesitant to modify the stuff from Icomoon, but... we really don't need the CSS file once we copy its contents here.

File: pattern-library/fonts/edx-icons/edx-icons.svg (hunk 9 above)
Comment (talbs) [inline pattern-library/fonts/edx-icons/edx-icons.svg]: Cool, it may make sense to add a canned comment in all files that are generated by IcoMoon (.svg, .css, etc.)

File: pattern-library/sass/components/_icons.scss (hunk 13 above)
Comment (talbs) [inline pattern-library/sass/components/_icons.scss]: Cool, so this value (``currentColor``) won't work then, right? If not, can you adjust it so that it inherits the color of the parent?

Comment (clrux) [inline pattern-library/sass/components/_icons.scss]: I believe `currentColor` still works.

**Comment by cptvitamin** [comment]
I threw everything I had at this and it works great.  :+1:

**Comment by clrux** [comment]
@talbs I left you a comment on [your other PR](https://github.com/edx/ux-pattern-library/pull/174#issuecomment-144597773).

@frrrances, @talbs You guys should continue your review despite the remaining warning. Please and thanks.

**Comment by talbs** [comment]
@clrux, thanks. Left you a comment (https://github.com/edx/ux-pattern-library/pull/174#issuecomment-144710016) back there. Let me know if you still have concerns.

Also, when you update the documentation in the Icons wiki page (and maybe for the description of this PR too), can you make sure you address the two major use cases of using an icon?

* Decorative - no semantic/content value is communicated through the icon
* UI enhancing/replacing - the icon has semantic/content value and should have a fallback for universally comprehension.

**Comment by clrux** [comment]
On it!

**Comment by clrux** [comment]
@talbs I've fixed the Modernizr script. I opted to generate a new (updated) version directly from Modernizr using the same checks (generatedcontent and fontface). The new version didn't add quite the same prefixes, so I had to update the A Font Garde scripts too, to recognize the ones that Modernizr adds. So all in all, this is a pretty customized package and updating from the vendor(s) won't really work. But, as you said, these aren't technically vendor packages so it shouldn't matter.

If you and @frrrances are okay with this, I'll squash things down. I'll probably get to documentation later and I won't merge until that's squared away.

**Comment by talbs** [comment]
@clrux, I'm getting the follow missing resource error locally when running this branch.

screen shot 2015-10-01 at 12 35 36 pm

Am I missing something or need to update anything locally?

**Comment by talbs** [comment]
^^ @clrux, I tried starting the default ``gulp`` task again and the icons rendered properly with no missing resources. Any reason why this would not have happened the first time for me?

**Comment by clrux** [comment]
Hmm, not getting that. Let me play a bit.

File: pattern-library/sass/components/_forms.scss [hunk 16]
```diff
@@ -287,7 +287,12 @@
         position: absolute;
         @include right(rem(($baseline/4)));
         margin-top: rem(-($baseline/10)); // vertically centers the down caret icon
-        fill: $input-control-color;
+        color: $input-control-color;
+        vertical-align: middle;
+    }
+
+    .icon-fallback-glyph .icon:before {
+        content: "\25BE"; // system glyph down arrow
```
Comment (talbs) [inline pattern-library/sass/components/_forms.scss]: Nice touch! Any way we can make this even easier for folks who need to manually add fall backs and create Sass vars for known/managed icons that we're falling back specifically? If so, I could see us using a var like ``icon-caret-down-fallback`` and then calling it here.

File: pldoc/static/js/pattern-library.js [hunk 17]
```diff
@@ -3,6 +3,9 @@ require.config({
     paths: {
         jquery: "/public/js/jquery.min",
         svg4everybody: "/public/js/svg4everybody.min",
+        modernizr: "/public/js/modernizr-custom.js",
```
Comment (talbs) [inline pldoc/static/js/pattern-library.js]: Nice touch re-naming this file to make it much clearer what's going on.

File: index.html (hunk 6 above)
Comment (talbs) [inline index.html]: Yeah, that's cool. One NIT: can we make the initial size of these icons larger. Before sliding, they seem smaller than they once were.

**Comment by talbs** [comment]
@clrux, this is looking good. A few loose ends:

* Sweep for any un-needed files in the repo or this PR (as its gone through a few directions and experiments)
* Try to reproduce my issue of not seeing icons/missing resources from a fresh start locally (gulp clean, npm install, bower install, gulp).
* DOC Updates - your order of steps sounds good, just make sure to cover the topics in this PR as well as really describe how to maintain icons (through IcoMoon) in a separate section. I want it to be clear how one would add/remove icons with this new adjustment

I'll try to walk @frrrances through this tomorrow. Once that's all done this has my :+1:

**Comment by clrux** [comment]
@talbs I've gone through and done some cleanup. I removed the SVG files because they weren't being loaded in our `_variables.scss` file and so we didn't need them.

I also believe I've fixed the error you were getting. I believe it was a race condition in which our `edx-icons.js` was loading before `afontgarde` so I just required the latter in the former.

I'm going to work on the documentation tonight and tomorrow so we can hopefully merge this by end of business tomorrow.

**Comment by talbs** [comment]
> @talbs I've gone through and done some cleanup. I removed the SVG files because they weren't being loaded in our _variables.scss file and so we didn't need them.

> I also believe I've fixed the error you were getting. I believe it was a race condition in which our edx-icons.js was loading before afontgarde so I just required the latter in the former.

> I'm going to work on the documentation tonight and tomorrow so we can hopefully merge this by end of business tomorrow.

:guitar: on to all of that! Thanks! I'll do a walkthrough tomorrow to get our ends all set.

**Comment by clrux** [comment]
@talbs @frrrances Wiki documentation updated! https://github.com/edx/ux-pattern-library/wiki/Styleguide:-Icons

Would love to merge this tomorrow!

**Comment by clrux** [comment]
Had issues trying to squash commits. I did `git rebase -i HEAD~22` but it kept showing like 50+ commits, and the merge commit wasn't included. Rather than risk messing with it, I decided to just make a new branch and cherry pick the work on to it.
